programs/bpf_loader/src/serialization.rs

```diff
@@ -67,22 +72,38 @@ impl Serializer {
         unsafe {
             self.buffer.write_unchecked(value);
         }
+
+        vaddr
     }
 
-    fn write_all(&mut self, value: &[u8]) {
+    fn write_all(&mut self, value: &[u8]) -> u64 {
+        let vaddr = self
+            .vaddr
+            .saturating_add(self.buffer.len() as u64)
+            .saturating_sub(self.region_start as u64);
         // Safety:
         // see write() - the buffer is guaranteed to be large enough
         unsafe {
             self.buffer.write_all_unchecked(value);
         }
+
+        vaddr
     }
 
-    fn write_account(&mut self, account: &mut BorrowedAccount<'_>) -> Result<(), InstructionError> {
-        if self.copy_account_data {
+    fn write_account(
+        &mut self,
+        account: &mut BorrowedAccount<'_>,
+    ) -> Result<u64, InstructionError> {
+        let vm_data_addr = if self.copy_account_data {
+            let vm_data_addr = self.vaddr.saturating_add(self.buffer.len() as u64);
```

Comment:
hmm, `Serializer` code is hard to reason about due to naming....

Specifically, lack of `.saturating_sub(self.regision_start)` unlike other `vaddr` calcs here can only be assured to be safe because `self.copy_account_data` codepath never issue `.push_region()` except via the finalization (`.finish()`).

imo, this is brittle. so, how about these some rearrangement?

```rust
struct Serializer {
    ...
    region_vaddr: u64, // renamed from vaddr
    region_buffer_{pos,offset}: u64, // renamed from region_start
    ...
}
 
impl Serializer {
   fn flush_region(...) { ... } // renamed from push_region
   fn add_account_data_region(...) {...} // renamed from push_account_data_region
   fn current_vaddr(...) { // new; extract to this for dry
       self
            .region_vaddr
            .saturating_add(self.buffer.len() as u64)
            .saturating_sub(self.region_start as u64);
   }
}
```

Reply:
`flush` vs `add` verbs are deliberate to differentiate the semantics of difference of semantically consuming `self.buffer` or not.

Reply:
@alessandrod maybe this and bunch of nits (https://github.com/solana-labs/solana/pull/31986#discussion_r1275807662, https://github.com/solana-labs/solana/pull/31986#discussion_r1275808696, https://github.com/solana-labs/solana/pull/31986#discussion_r1275814316) in this file will be addredded as a follow up pr?

if this is too bothersome for you, I'll create a follow up pr, assuming you like the general taste of these. :)